File: java/Exercises.java

import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.Optional;
import java.util.function.Predicate;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class Exercises {

    static Map<Integer, Long> change(long amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative");
        }
        var counts = new HashMap<Integer, Long>();
        for (var denomination : List.of(25, 10, 5, 1)) {
            counts.put(denomination, amount / denomination);
            amount %= denomination;
        }
        return counts;
    }

    public static Optional<String> firstThenLowerCase(List<String> strings, Predicate<String> predicate) {
        return strings.stream().filter(predicate).findFirst().map(String::toLowerCase);
    }

    static class Say {

        private String phrase;

        public Say(String phrase) {
            this.phrase = phrase;
        }

        public String phrase() {
            return this.phrase;
        }

        public Say and(String string) {
            return new Say(this.phrase + " " + string);
        }
    }

    public static Say say(String phrase) {
        return new Say(phrase);
    }

    public static Say say() {
        return new Say("");
    }

    //return not a int 
    static Integer meaningfulLineCount(String fileName) throws IOException {

        try (var fileReader = new BufferedReader(new FileReader(fileName))) {
            return (int) fileReader.lines().filter(line -> !line.trim().isEmpty() && !line.trim().startsWith("#")).count();
        }
    }
}

record Quaternion(double a, double b, double c, double d) {

    public Quaternion    {
        if (Double.isNaN(a) || Double.isNaN(b) || Double.isNaN(c) || Double.isNaN(d)) {
            throw new IllegalArgumentException("Coefficients cannot be NaN");
        }
    }

    public final static Quaternion ZERO = new Quaternion(0, 0, 0, 0);
    public final static Quaternion I = new Quaternion(0, 1, 0, 0);
    public final static Quaternion J = new Quaternion(0, 0, 1, 0);
    public final static Quaternion K = new Quaternion(0, 0, 0, 1);

    public Quaternion conjugate() {
        return new Quaternion(a, -b, -c, -d);
    }

    public List<Double> coefficients() {
        return List.of(a, b, c, d);
    }

    public Quaternion plus(Quaternion other) {

        return new Quaternion(
                this.a + other.a,
                this.b + other.b,
                this.c + other.c,
                this.d + other.d
        );

    }

    public Quaternion times(Quaternion other) {

        return new Quaternion(
                this.a * other.a - this.b * other.b - this.c * other.c - this.d * other.d,
                this.b * other.a + this.a * other.b + this.c * other.d - this.d * other.c,
                this.a * other.c - this.b * other.d + this.c * other.a + this.d * other.b,
                this.a * other.d + this.b * other.c - this.c * other.b + this.d * other.a
        );
    }

    @Override
    public String toString() {
        String s = "";
        var coefficients = this.coefficients();
        var terms = List.of("", "i", "j", "k");

        for (int i = 0; i < coefficients.size(); i++) {
            var coefficient = coefficients.get(i);

            if (coefficient == 0) {
                continue;
            }

            s += coefficient < 0 ? "-" : s.isEmpty() ? "" : "+";

            if (Math.abs(coefficient) != 1 || i == 0) {
                s += Math.abs(coefficient);
            }

            s += terms.get(i);
        }

        return s.length() > 0 ? s : "0";
    }

}

sealed interface BinarySearchTree permits Empty, Node {

    int size();

    boolean contains(String value);

    BinarySearchTree insert(String value);

}

final record Empty() implements BinarySearchTree {

    @Override
    public int size() {
        return 0;
    }

    @Override
    public boolean contains(String value) {
        return false;
    }

    @Override
    public BinarySearchTree insert(String value) {
        return new Node(value, this, this);
    }

    @Override
    public String toString() {
        return "()";
    }
}

final class Node implements BinarySearchTree {

    private final String value;
    private final BinarySearchTree left;
    private final BinarySearchTree right;

    Node(String value, BinarySearchTree left, BinarySearchTree right) {
        this.value = value;
        this.left = left;
        this.right = right;
    }

    @Override
    public int size() {
        return 1 + left.size() + right.size();
    }

    @Override
    public boolean contains(String value) {
        return this.value.equals(value) || left.contains(value) || right.contains(value);
    }

    @Override
    public BinarySearchTree insert(String value) {
        if (value.compareTo(this.value) < 0) {
            return new Node(this.value, left.insert(value), right);
        } else {
            return new Node(this.value, left, right.insert(value));
        }
    }

    @Override
    public String toString() {
        var stringRep = "(" + left + value + right + ")";
        return stringRep.replace("()", "");

    }
}

File: kotlin/Exercises.kt
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException
import kotlin.math.abs


fun change(amount: Long): Map<Int, Long> {
    require(amount >= 0) { "Amount cannot be negative" }
    
    val counts = mutableMapOf<Int, Long>()
    var remaining = amount
    for (denomination in listOf(25, 10, 5, 1)) {
        counts[denomination] = remaining / denomination
        remaining %= denomination
    }
    return counts
}

fun firstThenLowerCase(strings: List<String>, predicate: (String) -> Boolean): String? {
    return strings.firstOrNull(predicate)?.lowercase() 
}

data class Say(val phrase: String) {
    fun and(nextPhrase: String): Say {
        return Say("$phrase $nextPhrase")
    }
}

fun say(phrase: String = ""): Say {
    return Say(phrase)
}
@Throws(IOException::class)
fun meaningfulLineCount(fileName: String): Long {
    BufferedReader(FileReader(fileName)).use { 
        reader -> return reader.useLines{lines -> lines.filter {it.trim().isNotEmpty() && (it.trim().firstOrNull() != '#')}.count().toLong()
        }
    }
}

data class Quaternion(val a: Double, val b: Double, val c: Double, val d: Double){
    
    companion object {
        val ZERO = Quaternion(0.0, 0.0, 0.0, 0.0)
        val I = Quaternion(0.0, 1.0, 0.0, 0.0)
        val J = Quaternion(0.0, 0.0, 1.0, 0.0)
        val K = Quaternion(0.0, 0.0, 0.0, 1.0)
    }
    
    operator fun plus(other: Quaternion): Quaternion {
        return Quaternion(a + other.a, b + other.b, c + other.c, d + other.d)
    }

    operator fun times(other: Quaternion): Quaternion{
        return Quaternion(
                a * other.a - b * other.b - c * other.c - d * other.d,
                b * other.a + a * other.b + c * other.d - d * other.c,
                a * other.c - b * other.d + c * other.a + d * other.b,
                a * other.d + b * other.c - c * other.b + d * other.a
        )
    }

    fun coefficients(): List<Double> = listOf(a, b, c, d)

    fun conjugate(): Quaternion = Quaternion(a, -b, -c, -d)

    override fun toString(): String {
        var s = ""
        var coefficients = this.coefficients()
        var terms = listOf("", "i", "j", "k")

        for ((index, coefficient) in coefficients.withIndex()) {
            if (coefficient == 0.0) {
                continue
            }

            s += if (coefficient < 0) "-" else if (s.isEmpty()) "" else "+"

            if (abs(coefficient) != 1.0 || index == 0) {
                s += abs(coefficient)
            }

            s += terms.get(index)
        }

        return if (s.length > 0) s else "0"

    }
}

sealed interface BinarySearchTree {
    fun size(): Int
    fun contains(value: String): Boolean
    fun insert(value: String): BinarySearchTree

    object Empty: BinarySearchTree {
        override fun size(): Int = 0
        override fun contains(value: String): Boolean = false
        override fun insert(value: String): BinarySearchTree = Node(value, Empty, Empty)
        override fun toString(): String = "()"
    }

    data class Node(private val value: String, private val left: BinarySearchTree, private val right: BinarySearchTree): BinarySearchTree {
        override fun size(): Int {
            return 1 + right.size() + left.size()
        }
        override fun contains(value: String): Boolean = when {
            value <  this.value -> left.contains(value)
            value > this.value -> right.contains(value)
            else -> true
        }
        override fun insert(value: String): BinarySearchTree = when {
            value < this.value -> Node(this.value, left.insert(value), right)
            value > this.value -> Node(this.value, left, right.insert(value))
            else -> this
        }
        override fun toString(): String {
            return ("(" + left + value + right + ")").replace("()", "")
        }
    }

}

